core/src/main/scala/org/apache/spark/storage/DiskBlockManager.scala

```diff
@@ -46,18 +46,35 @@ private[spark] class DiskBlockManager(conf: SparkConf, deleteFilesOnStop: Boolea
     System.exit(ExecutorExitCode.DISK_STORE_FAILED_TO_CREATE_DIR)
   }
 
+  private def containerDirEnabled: Boolean = Utils.isRunningInYarnContainer(conf)
+
+  /* Create container directories on YARN to persist the temporary files.
+   * (temp_local, temp_shuffle)
+   * These files have no opportunity to be cleaned before application end on YARN.
+   * This is a real issue, especially for long-lived Spark application like Spark thrift-server.
+   * So we persist these files in YARN container directories which could be cleaned by YARN when
+   * the container exists. */
+  private[spark] val containerDirs: Array[File] =
+    if (containerDirEnabled) createContainerDirs(conf) else Array.empty[File]
+
   private[spark] val localDirsString: Array[String] = localDirs.map(_.toString)
 
   // The content of subDirs is immutable but the content of subDirs(i) is mutable. And the content
   // of subDirs(i) is protected by the lock of subDirs(i)
   private val subDirs = Array.fill(localDirs.length)(new Array[File](subDirsPerLocalDir))
 
+  private val subContainerDirs = if (containerDirEnabled) {
+    Array.fill(containerDirs.length)(new Array[File](subDirsPerLocalDir))
+  } else {
+    Array.empty[Array[File]]
+  }
+
   private val shutdownHook = addShutdownHook()
 
-  /** Looks up a file by hashing it into one of our local subdirectories. */
   // This method should be kept in sync with
   // org.apache.spark.network.shuffle.ExecutorDiskUtils#getFile().
```

Comment:
@LantaoJin . This comment reminds me that we need to handle `ExecutorDiskUtils` properly. Do we need to update `ExecutorDiskUtils.getFile`? Or, do we need to remove this comment?

Reply:
No. I didn't change any behaviour for non-temp files. This `ExecutorDiskUtils.getFile` is handing the non-temp files such as start with "shuffle_", "rdd_". So it should still be kept in sync with `DiskBlockManager.getFile`.

Reply:
You can see the method body is not changed.

Reply:
Why do we need to have unused parameters?

Reply:
```scala
  def getFile(blockId: BlockId): File = {
    if (containerDirEnabled && blockId.isTemp) {
      getFile(containerDirs, subContainerDirs, subDirsPerLocalDir, blockId.name)
    } else {
      getFile(localDirs, subDirs, subDirsPerLocalDir, blockId.name)
    }
  }
```
Now we just storage temp files and non-temp files to different root paths. The algorithm how to find a file doesn't change.

Reply:
> Why does this function need to have unused parameters?
Which parameters?

Reply:
> Why does this function need to have unused parameters?

Ah, you must mean the `getFile` in `DiskBlockManager` has 4 parameters and `getFile` in `ExecutorDiskUtils` has 3 parameters. ExecutorDiskUtils is a utility, so 3 parameters can define a file path structure. The parameter `subDirs` in `DiskBlockManager` only is a local variable which used in other parts. It does not impact the finding algorithm result.

Reply:
So the `getFile` in `DiskBlockManager` has 4 parameters:
```scala
private def getFile(localDirs: Array[File], subDirs: Array[Array[File]],
      subDirsPerLocalDir: Int, filename: String): File
```

Reply:
I think we don't need to pay more attention on how parameters in `getFile`. Before this PR. There was only 1 parameter `filename: String` in DiskBlockManager and 3 parameters in `ExecutorDiskUtils`.